include/boost/url/grammar/is_mutable_string.hpp

```diff
@@ -0,0 +1,78 @@
+//
+// Copyright (c) 2016-2019 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_IS_MUTABLE_STRING_HPP
+#define BOOST_URL_IS_MUTABLE_STRING_HPP
+
+#include <type_traits>
+
+namespace boost {
+namespace urls {
+namespace grammar {
+
+/** Alias for `std::true_type` if `T` satisfies __MutableString__
+    for the iterator type `I`
+*/
+#ifdef BOOST_URL_DOCS
+template<class T, class I>
+using is_mutable_string = __see_below__;
+#else
+namespace detail
+{
+template<class T, class = void>
+struct has_value_type : std::false_type {};
+
+template<class T>
+struct has_value_type<T, boost::void_t<
+    typename T::value_type>>
+    : std::true_type
+{};
+
+template<class T, class I, class = void>
+struct has_assign_and_append : std::false_type {};
+
+template<class T, class I>
+struct has_assign_and_append<T, I, boost::void_t<
+    // T::append(I, I)
+    decltype(
+        std::declval<T&>()
+            .append(
+                std::declval<I>(),
+                std::declval<I>())
+                ),
+    // T::assign(I, I)
+    decltype(
+        std::declval<T&>()
+            .assign(
+                std::declval<I>(),
+                std::declval<I>())
+                )>>
+    : std::true_type
+{};
+}
+
+template <class T, class I, class = void>
+struct is_mutable_string : std::false_type {};
+
+template <class T, class I>
+struct is_mutable_string<
+    T, I,
+    typename std::enable_if<
+        detail::has_value_type<T>::value &&
+        detail::has_assign_and_append<T, I>::value
+        >::type>
+    : std::is_same<typename T::value_type, char>
```

Comment:
Oh you can do better than this :) do it all in `is_mutable_string`

Reply:
Everything was in `is_mutable_string`, but I didn't manage to include the `value_type` requirement without splitting it.